### PR TITLE
Remove 'kind' from being part of a checksum

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Experiment to see what breaks.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs [hunk 1]
```diff
@@ -87,42 +87,42 @@ IEnumerator IEnumerable.GetEnumerator()
...
         public ProjectReferenceChecksumCollection(Checksum[] checksums) : this((object[])checksums) { }
-        public ProjectReferenceChecksumCollection(object[] checksums) : base(WellKnownSynchronizationKind.ProjectReferenceChecksumCollection, checksums) { }
+        public ProjectReferenceChecksumCollection(object[] checksums) : base(checksums) { }
     }
 
     internal class MetadataReferenceChecksumCollection : ChecksumCollection
     {
         public MetadataReferenceChecksumCollection(Checksum[] checksums) : this((object[])checksums) { }
-        public MetadataReferenceChecksumCollection(object[] checksums) : base(WellKnownSynchronizationKind.MetadataReferenceChecksumCollection, checksums) { }
+        public MetadataReferenceChecksumCollection(object[] checksums) : base(checksums) { }
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs:120]: 💭 For this case, the type seems desirable.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs:120]: how come?  afaict, nothing reads these.

File: src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs [hunk 2]
```diff
@@ -82,47 +82,4 @@ IEnumerator IEnumerable.GetEnumerator()
...
-    {
-        public MetadataReferenceChecksumCollection(Checksum[] checksums) : this((object[])checksums) { }
-        public MetadataReferenceChecksumCollection(object[] checksums) : base(WellKnownSynchronizationKind.MetadataReferenceChecksumCollection, checksums) { }
-    }
-
-    internal class AnalyzerReferenceChecksumCollection : ChecksumCollection
-    {
-        public AnalyzerReferenceChecksumCollection(Checksum[] checksums) : this((object[])checksums) { }
-        public AnalyzerReferenceChecksumCollection(object[] checksums) : base(WellKnownSynchronizationKind.AnalyzerReferenceChecksumCollection, checksums) { }
-    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs:127]: we could consider still having these, just for the slight amount of strong type safety it affords in one locatin.  but it seemed pretty overkill to me.

**Comment by CyrusNajmabadi** [comment]
@sharwell for thoughts on this.

File: src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs [hunk 3]
```diff
@@ -17,13 +17,13 @@ namespace Microsoft.CodeAnalysis.Serialization
     /// <summary>
     /// collection which children is checksum.
     /// </summary>
-    internal abstract class ChecksumCollection : ChecksumWithChildren, IEnumerable<Checksum>
+    internal class ChecksumCollection : ChecksumWithChildren, IEnumerable<Checksum>
```
Comment (tmat) [inline src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs:20]: sealed?

**Review by tmat — APPROVED** [review]

**Comment by tmat** [comment]
@CyrusNajmabadi Nice! Can we remove the other kinds as well? Doesn't the reader always know the expected structure of the data?

File: src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/ChecksumCollection.cs:20]: yup.  seems reasonable :)

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Nice! Can we remove the other kinds as well? Doesn't the reader always know the expected structure of the data?

@tmat great question.  i'll look into this.  if we can just make it a simple reader/writer with no need for this info, that would be great!

**Comment by tmat** [comment]
Yeah, I was wondering about that a while ago, but didn't have time to follow up. Thanks for cleaning this up!

**Pull request merged**